File: .zo/scripts/python/setup-specify-idea.py
#!/usr/bin/env python3
"""
Setup for Spec-Driven Development - Find brainstorm file and spec template.

This script locates the appropriate brainstorm file and outputs paths needed
for creating a spec from an idea. It supports both JSON and plain text output.

Usage:
    python3 setup-specify-idea.py [--json] [brainstorm_file]
    
Arguments:
    --json, -h    Output results in JSON format
    brainstorm_file  Optional path to brainstorm file (absolute or relative)

The script searches for brainstorm files in the following order:
1. User-provided path (if specified)
2. .zo/brainstorms/ (primary location)
3. FEATURE_DIR/brainstorms/ (legacy)
4. docs/brainstorms/ (legacy)

Example:
    python3 setup-specify-idea.py
    python3 setup-specify-idea.py --json
    python3 setup-specify-idea.py path/to/brainstorm.md
"""

import argparse
import json
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))
from python.common import get_repo_root, get_feature_dir


def find_brainstorm_file(repo_root: str, feature_dir: str, user_path: str = None) -> Path:
    """
    Find the brainstorm file from various possible locations.
    
    Args:
        repo_root: Repository root directory
        feature_dir: Current feature directory
        user_path: Optional user-provided path
        
    Returns:
        Path to the brainstorm file
        
    Raises:
        FileNotFoundError: If no brainstorm file is found
    """
    if user_path:
        # User provided a path - check if it exists
        user_path_obj = Path(user_path)
        if not user_path_obj.is_absolute():
            # Relative path - resolve from current working directory
            user_path_obj = Path.cwd() / user_path_obj
        
        if user_path_obj.exists() and user_path_obj.is_file():
            return user_path_obj.resolve()
        else:
            raise FileNotFoundError(f"File '{user_path}' not found")
    
    # Auto-detect latest brainstorm file
    
    # Primary Location: .zo/brainstorms
    primary_dir = Path(repo_root) / '.zo' / 'brainstorms'
    if primary_dir.exists():
        # Find all .md files and sort by modification time (most recent first)
        md_files = list(primary_dir.glob('*.md'))
        if md_files:
            latest_file = max(md_files, key=lambda p: p.stat().st_mtime)
            return latest_file
    
    # Legacy location 1: FEATURE_DIR/brainstorms
    legacy_dir_1 = Path(feature_dir) / 'brainstorms'
    if legacy_dir_1.exists():
        md_files = list(legacy_dir_1.glob('brainstorm-*.md'))
        if md_files:
            latest_file = max(md_files, key=lambda p: p.stat().st_mtime)
            return latest_file
    
    # Legacy location 2: docs/brainstorms
    legacy_dir_2 = Path(repo_root) / 'docs' / 'brainstorms'
    if legacy_dir_2.exists():
        md_files = list(legacy_dir_2.glob('brainstorm-*.md'))
        if md_files:
            latest_file = max(md_files, key=lambda p: p.stat().st_mtime)
            return latest_file
    
    raise FileNotFoundError(
        "No brainstorm file found. Run 'zo.brainstorm' first."
    )


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description='Find brainstorm file and spec template for spec creation',
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.add_argument(
        '--json',
        action='store_true',
        help='Output results in JSON format'
    )
    parser.add_argument(
        'brainstorm_file',
        nargs='?',
        help='Optional path to brainstorm file (absolute or relative)'
    )
    
    args = parser.parse_args()
    
    try:
        # Get repository paths
        repo_root = get_repo_root()
        current_branch = Path(get_feature_dir(repo_root, get_repo_root())).name
        feature_dir = get_feature_dir(repo_root, current_branch)
        
        # Find brainstorm file
        brainstorm_file = find_brainstorm_file(
            repo_root,
            feature_dir,
            args.brainstorm_file
        )
        
        # Define other paths
        spec_template = Path(repo_root) / '.zo' / 'templates' / 'spec-from-idea.md'
        design_file = Path(feature_dir) / 'design.md'
        
        # Output results - use compact JSON to match bash
        if args.json:
            result = {
                'BRAINSTORM_FILE': str(brainstorm_file),
                'SPEC_TEMPLATE': str(spec_template),
                'DESIGN_FILE': str(design_file)
            }
            print(json.dumps(result, separators=(',', ':')))
        else:
            print(f"BRAINSTORM_FILE: {brainstorm_file}")
            print(f"SPEC_TEMPLATE: {spec_template}")
            print(f"DESIGN_FILE: {design_file}")
        
        return 0
        
    except FileNotFoundError as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: .roo/tools/setup-specify-idea.ts
/**
 * setup_specify_idea - Find brainstorm file and spec template for spec creation
 *
 * Locates the appropriate brainstorm file and outputs paths needed for
 * creating a spec from an idea. Searches multiple locations for brainstorm files
 * in priority order: user-provided path, .zo/brainstorms/, FEATURE_DIR/brainstorms/, docs/brainstorms/
 *
 * Ported from: .zo/scripts/python/setup-specify-idea.py
 *
 * @example
 * ```typescript
 * const result = await setup_specify_idea({ json: true })
 * // Returns: {"BRAINSTORM_FILE": "/path/to/brainstorms/idea-2025-01-14.md", "SPEC_TEMPLATE": "/path/to/templates/spec-from-idea.md"}
 * ```
 *
 * @param json - Output results in JSON format (default: false)
 * @param brainstormFile - Optional path to brainstorm file (absolute or relative)
 *
 * @returns JSON object or formatted text with brainstorm file and template paths
 *
 * @throws {Error} If execution environment validation fails
 * @throws {Error} If no brainstorm file is found
 * @throws {Error} If user-provided path doesn't exist
 *
 * @see {@link ../lib/util.ts} for path resolution utilities
 */

import { defineCustomTool } from '@roo-code/types';
import { parametersSchema as z } from '@roo-code/types';
import { existsSync, readdirSync, statSync } from 'node:fs';
import { join, resolve } from 'node:path';
import { cwd } from 'node:process';
import {
  getFeatureDir,
  getRepoRoot,
  validateExecutionEnvironment,
} from './lib/util.js';

export default defineCustomTool({
  name: 'setup_specify_idea',

  description: `
    Find brainstorm file and spec template for creating a spec from an idea.
    Searches for brainstorm files in .zo/brainstorms/, FEATURE_DIR/brainstorms/, or docs/brainstorms/.
    Returns paths needed for spec creation workflow.
  `,

  parameters: z.object({
    json: z
      .boolean()
      .optional()
      .describe('Output results in JSON format (compact)'),
    brainstormFile: z
      .string()
      .optional()
      .describe('Optional path to brainstorm file (absolute or relative)'),
  }),

  async execute({ json = false, brainstormFile }, _context) {
    try {
      // Validate execution environment
      if (!validateExecutionEnvironment()) {
        const errorMsg = 'ERROR: Execution environment validation failed.';
        return json
          ? JSON.stringify({ error: errorMsg })
          : errorMsg;
      }

      // Get repository paths
      const repoRoot = getRepoRoot();
      const currentBranch = getCurrentBranch();
      const featureDir = getFeatureDir(repoRoot, currentBranch);

      // Find brainstorm file
      let actualBrainstormFile: string;
      try {
        actualBrainstormFile = findBrainstormFile(repoRoot, featureDir, brainstormFile);
      } catch (error) {
        const errorMsg = `Error: ${(error as Error).message}`;
        return json
          ? JSON.stringify({ error: errorMsg })
          : errorMsg;
      }

      // Define other paths
      const specTemplate = join(repoRoot, '.zo', 'templates', 'spec-from-idea.md');
      const designFile = join(featureDir, 'design.md');

      // Prepare result
      const result = {
        BRAINSTORM_FILE: actualBrainstormFile,
        SPEC_TEMPLATE: specTemplate,
        DESIGN_FILE: designFile,
      };

      // Return in requested format
      if (json) {
        return JSON.stringify(result);
      }

      // Formatted text output
      return [
        `BRAINSTORM_FILE: ${result.BRAINSTORM_FILE}`,
        `SPEC_TEMPLATE: ${result.SPEC_TEMPLATE}`,
        `DESIGN_FILE: ${result.DESIGN_FILE}`,
      ].join('\n');

    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return json
        ? JSON.stringify({ error: errorMessage })
        : `Error: ${errorMessage}`;
    }
  },
});

/**
 * Get current branch (simplified version)
 */
function getCurrentBranch(): string {
  // Try to get from git
  try {
    const { execSync } = require('node:child_process');
    const result = execSync('git rev-parse --abbrev-ref HEAD', {
      encoding: 'utf-8',
      stdio: ['ignore', 'pipe', 'ignore'],
    });
    return result.trim();
  } catch {
    // Fallback to feature directory name
    return cwd().split('/').pop() || 'main';
  }
}

/**
 * Find the brainstorm file from various possible locations
 *
 * Searches in the following order:
 * 1. User-provided path (if specified)
 * 2. .zo/brainstorms/ (primary location)
 * 3. FEATURE_DIR/brainstorms/ (legacy)
 * 4. docs/brainstorms/ (legacy)
 */
function findBrainstormFile(repoRoot: string, featureDir: string, userPath?: string): string {
  if (userPath) {
    // User provided a path - check if it exists
    const userPathObj = resolve(userPath);
    if (existsSync(userPathObj) && statSync(userPathObj).isFile()) {
      return userPathObj;
    }
    // Try relative to current directory
    const cwdRelative = resolve(cwd(), userPath);
    if (existsSync(cwdRelative) && statSync(cwdRelative).isFile()) {
      return cwdRelative;
    }
    throw new Error(`File '${userPath}' not found.`);
  }

  // Auto-detect latest brainstorm file

  // Primary Location: .zo/brainstorms
  const primaryDir = resolve(repoRoot, '.zo', 'brainstorms');
  if (existsSync(primaryDir)) {
    const latestFile = findLatestMdFile(primaryDir);
    if (latestFile) {
      return latestFile;
    }
  }

  // Legacy location 1: FEATURE_DIR/brainstorms
  const legacyDir1 = resolve(featureDir, 'brainstorms');
  if (existsSync(legacyDir1)) {
    const latestFile = findLatestMdFile(legacyDir1, 'brainstorm-');
    if (latestFile) {
      return latestFile;
    }
  }

  // Legacy location 2: docs/brainstorms
  const legacyDir2 = resolve(repoRoot, 'docs', 'brainstorms');
  if (existsSync(legacyDir2)) {
    const latestFile = findLatestMdFile(legacyDir2, 'brainstorm-');
    if (latestFile) {
      return latestFile;
    }
  }

  throw new Error(
    'No brainstorm file found. Run \'zo.brainstorm\' first.'
  );
}

/**
 * Find the latest .md file in a directory by modification time
 */
function findLatestMdFile(dir: string, prefix: string = ''): string | null {
  try {
    const entries = readdirSync(dir, { withFileTypes: true });
    const mdFiles: Array<{ path: string; mtime: number }> = [];

    for (const entry of entries) {
      if (entry.isFile() && entry.name.endsWith('.md')) {
        if (prefix && !entry.name.startsWith(prefix)) {
          continue;
        }
        const fullPath = resolve(dir, entry.name);
        const stats = statSync(fullPath);
        mdFiles.push({ path: fullPath, mtime: stats.mtimeMs });
      }
    }

    if (mdFiles.length === 0) {
      return null;
    }

    // Sort by modification time (most recent first)
    mdFiles.sort((a, b) => b.mtime - a.mtime);
    return mdFiles[0]!.path;
  } catch {
    return null;
  }
}
